# JSON and XML - Lab

## Introduction

In this lab, you'll practice navigating JSON and XML data structures.

## Objectives
You will be able to:
* Effectively use the JSON module to load and parse JSON documents
* Read and access data stored in JSON and XML
* Compare  and contrast the JSON and XML as data interchange types


## XML

In [1]:
import xml.etree.ElementTree as ET

### Create an XML tree and retrieve the root tag.

In [3]:
ls

2001_Campaign_Contributions.csv*  LICENSE.md*                      README.md
CONTRIBUTING.md*                  nyc_2001_campaign_finance.json*
index.ipynb                       nyc_2001_campaign_finance.xml


In [4]:
#Your code here
tree = ET.parse('nyc_2001_campaign_finance.xml')
root = tree.getroot()

### How many direct descendents does the root tag have?

In [17]:
#Answer: 1
g = []
# for child in root:
#     for gchild in child:
#         g.append((gchild.tag, gchild.attrib))

for element in root.iter():
    g.append(element.tag)
    
g

['response',
 'row',
 'row',
 'candid',
 'candname',
 'officeboro',
 'canclass',
 'row',
 'election',
 'candid',
 'candname',
 'officecd',
 'officedist',
 'canclass',
 'primarypay',
 'generalpay',
 'runoffpay',
 'totalpay',
 'row',
 'election',
 'candid',
 'candname',
 'officecd',
 'officedist',
 'canclass',
 'primarypay',
 'generalpay',
 'runoffpay',
 'totalpay',
 'row',
 'election',
 'candid',
 'candname',
 'officecd',
 'officedist',
 'canclass',
 'primarypay',
 'generalpay',
 'runoffpay',
 'totalpay',
 'row',
 'election',
 'candid',
 'candname',
 'officecd',
 'officedist',
 'canclass',
 'primarypay',
 'generalpay',
 'runoffpay',
 'totalpay',
 'row',
 'election',
 'candid',
 'candname',
 'officecd',
 'officedist',
 'canclass',
 'primarypay',
 'generalpay',
 'runoffpay',
 'totalpay',
 'row',
 'election',
 'candid',
 'candname',
 'officecd',
 'officedist',
 'canclass',
 'primarypay',
 'generalpay',
 'runoffpay',
 'totalpay',
 'row',
 'election',
 'candid',
 'candname',
 'officecd',
 'o

### How many different types of tags are there within the entire XML file?

In [23]:
# Your code here
gset = list(set(g))
counter = [g.count(x) for x in gset]
tags = {}
for i in range(len(gset)):
    tags[g[i]] = counter[i]
tags

{'response': 285,
 'row': 284,
 'candid': 286,
 'candname': 284,
 'officeboro': 285,
 'canclass': 284,
 'election': 21,
 'officecd': 284,
 'officedist': 285}

### Create a DataFrame listing the number of each type of tag. 
Sort the DataFrame in descending order by the tag count. The first entry should demonstrate there are 286 row tags in the XML file.   
(Your DataFrame will be a single column, so could also be thought of as a Series.)

In [24]:
import pandas as pd

In [1]:
#Your code here
dfs = []
for n, element in enumerate(root.iter('row')):
    if n > 0:
        dfs.append(pd.DataFrame.from_dict(element.attrib, orient='index').transpose())
df = pd.concat(dfs)
print(len(df))
df.head()

NameError: name 'root' is not defined

## JSON

### Open the same dataset from json

In [ ]:
#Your code here

### What is the root data type of the json file?

In [ ]:
### Your code here

### Navigate to the 'data' key of your loaded json object. What data type is this?

In [ ]:
#Your code here

### Preview the first entry from the value returned by the 'data' key above.

In [ ]:
#Your code here

### Preview the Entry under meta -> view -> columns (the keys of three successively nested dictionaries)

### Create a DataFrame from your json data
The previous two questions previewed one entry from the data object within the json file, as well as the column details associated with that data from the meta entry within the json file. Both should have 19 entries. Create a DataFrame of the data. Be sure to use the information from the meta entry to add appropriate column names to your DataFrame.

In [ ]:
#Your code here

### What's wrong with the first row of the DataFrame?

In [ ]:
#Your code here

#Your answer here

## Summary

Congratulations! You've started exploring some more complicated data structures used for the web and got to practice data munging and exploring!